responseobjects/elastic_request_builder.py

```diff
@@ -258,16 +280,62 @@ def generate_paging_dict(es_response, pagination):
         :return: Modifies and returns the pagination updated with the
         new required entries.
         """
-        page_field = {
-            'count': len(es_response['hits']['hits']),
-            'total': es_response['hits']['total'],
-            'size': pagination['size'],
-            'from': pagination['from'],
-            'page': ((pagination['from'] - 1) / pagination['size']) + 1,
-            'pages': -(-es_response['hits']['total'] // pagination['size']),
-            'sort': pagination['sort'],
-            'order': pagination['order']
-        }
+        pages = -(-es_response['hits']['total'] // pagination['size'])
+        if es_response['hits']['total'] < SEARCH_AFTER_THRESHOLD:
+            # Use from/to pagination
+            page_field = {
+                'count': len(es_response['hits']['hits']),
+                'total': es_response['hits']['total'],
+                'size': pagination['size'],
+                'from': pagination['from'],
+                'page': ((pagination['from'] - 1) / pagination['size']) + 1,
+                'pages': pages,
+                'sort': pagination['sort'],
+                'order': pagination['order']
+            }
+        else:
+            # ...else use search_after/search_before pagination
+            es_hits = es_response['hits']['hits']
+            count = len(es_hits)
+
+            logging.debug("count="+str(count)+" and size="+str(pagination['size']))
+
+            if 'search_before' in pagination:
+                # hits are reverse sorted
+                if count > pagination['size']:
```

Comment:
@jvlasblom if the returned page length `count` (`es_response['hits']['hits']`) is restricted to `0<=count <=pagination['size']`, is there ever the possibility of entering this `if` statement? Does `search_after` append an extra array to `es_response['hits']['hits']`or something like that?

Reply:
I fetch one more result than the client requests (line 267), so that I can determine if there is a next page.  If the user requests a page of 10 results, I try to fetch 11.  If elasticsearch returns back 11, I give the client the first 10 results, and I know there is a 'next' page, otherwise I know it's the last page.  

It works the same for the 'previous' page, (essentially the same as a search_after pagination but with results reverse sorted, which end up being reversed again before I return them to the client -- necessary as elastic has no 'search_before').

Reply:
Completely missed that! Okay this is good.